panic: interface conversion: *v1.ExportTraceServiceRequest is not protoreflect.ProtoMessage: missing method ProtoReflect

**Describe the bug**
Tempo crashes when receiving OLTP telemetry

**To Reproduce**
Steps to reproduce the behavior:
1. Start Tempo via dockerhub image, "grafana/tempo:latest" at time of writing. `grafana/tempo@sha256:d6d1497d2b41ec74451dca762698c0321e3729b4bc83245f8b5c6360f481b625`
 ` tempo, version  (branch: main, revision: 045ede7f)`
Config:
```yaml
distributor:
  receivers:
    otlp:
      protocols:
        grpc:
```

2. Send some telemetry using Python library, using exact example provided on https://opentelemetry-python.readthedocs.io/en/stable/exporter/otlp/otlp.html
3. Tempo crashes:
```
panic: interface conversion: *v1.ExportTraceServiceRequest is not protoreflect.ProtoMessage: missing method ProtoReflect

goroutine 423 [running]:
github.com/grafana/tempo/pkg/gogocodec.(*gogoCodec).Unmarshal(0x1e89f00, {0xc000335800, 0x3c2, 0x3c2}, {0x212c5e0, 0xc000b1a300})
	/drone/src/pkg/gogocodec/gogocodec.go:65 +0xb3
google.golang.org/grpc.(*Server).processUnaryRPC.func2({0x212c5e0, 0xc000b1a300})
	/drone/src/vendor/google.golang.org/grpc/server.go:1263 +0x1f2
go.opentelemetry.io/collector/model/internal/data/protogen/collector/trace/v1._TraceService_Export_Handler({0x1e5f700, 0xc0009be740}, {0x27dc7c0, 0xc000b00a50}, 0xc0004d69c0, 0xc0009e60a0)
	/drone/src/vendor/go.opentelemetry.io/collector/model/internal/data/protogen/collector/trace/v1/trace_service.pb.go:206 +0x63
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0009fa000, {0x2809920, 0xc00017e600}, 0xc000532000, 0xc0009c09c0, 0x3af6660, 0x0)
	/drone/src/vendor/google.golang.org/grpc/server.go:1286 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc0009fa000, {0x2809920, 0xc00017e600}, 0xc000532000, 0x0)
	/drone/src/vendor/google.golang.org/grpc/server.go:1609 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/drone/src/vendor/google.golang.org/grpc/server.go:934 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/drone/src/vendor/google.golang.org/grpc/server.go:932 +0x294

```

**Expected behavior**
A: Tempo does not crash, even if the data provided is invalid. (I don't know enough about the opentelemetry-oython nor the wire protocol to tell if it is or not)
B: Tempo accepts the telemetry provided, assuming it was valid.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 * [ENHACEMENT] Enterprise jsonnet: add config to create tokengen job explicitly [#1256](https://github.com/grafana/tempo/pull/1256) (@kvrhdn)
 * [BUGFIX]: Remove unnecessary PersistentVolumeClaim [#1245](https://github.com/grafana/tempo/issues/1245)
 * [BUGFIX] Fixed issue when query-frontend doesn't log request details when request is cancelled [#1136](https://github.com/grafana/tempo/issues/1136) (@adityapwr)
+* [BUGFIX] Fixed panic when using etcd as ring's kvstore [#1260](https://github.com/grafana/tempo/pull/1260) (@mapno)
 
 
 ## v1.3.0 / 2022-01-24

--- a/cmd/tempo-serverless/go.mod
+++ b/cmd/tempo-serverless/go.mod
@@ -14,7 +14,6 @@ require (
 require (
 	cloud.google.com/go/functions v1.0.0 // indirect
 	cloud.google.com/go/kms v1.1.0 // indirect
-	github.com/cortexproject/cortex v1.10.1-0.20211214110443-07d6cfc4eb84 // indirect
 	github.com/google/uuid v1.3.0
 	github.com/grafana/dskit v0.0.0-20220112093026-95274ccc858d
 	github.com/weaveworks/common v0.0.0-20210913144402-035033b78a78

--- a/go.mod
+++ b/go.mod
@@ -258,7 +258,7 @@ require (
 	github.com/yuin/gopher-lua v0.0.0-20200816102855-ee81675732da // indirect
 	go.etcd.io/bbolt v1.3.6 // indirect
 	go.etcd.io/etcd v3.3.25+incompatible // indirect
-	go.etcd.io/etcd/api/v3 v3.5.0 // indirect
+	go.etcd.io/etcd/api/v3 v3.5.1 // indirect
 	go.etcd.io/etcd/client/pkg/v3 v3.5.0 // indirect
 	go.etcd.io/etcd/client/v3 v3.5.0 // indirect
 	go.mongodb.org/mongo-driver v1.7.5 // indirect

--- a/integration/e2e/config-microservices.tmpl.yaml
+++ b/integration/e2e/config-microservices.tmpl.yaml
@@ -12,8 +12,7 @@ distributor:
 ingester:
   lifecycler:
     ring:
-      kvstore:
-        store: memberlist
+      kvstore: {{ .KVStore }}
       replication_factor: 3
     heartbeat_period: 100ms
   trace_idle_period: 1s

--- a/integration/e2e/e2e_test.go
+++ b/integration/e2e/e2e_test.go
@@ -1,16 +1,20 @@
 package e2e
 
 import (
+	"bytes"
+	"fmt"
 	"math/rand"
 	"os"
+	"path/filepath"
 	"reflect"
 	"strings"
 	"sync"
 	"testing"
+	"text/template"
 	"time"
 
 	"github.com/grafana/e2e"
-	e2e_db "github.com/grafana/e2e/db"
+	e2edb "github.com/grafana/e2e/db"
 	jaeger_grpc "github.com/jaegertracing/jaeger/cmd/agent/app/reporter/grpc"
 	thrift "github.com/jaegertracing/jaeger/thrift-gen/jaeger"
 	"github.com/prometheus/prometheus/model/labels"
@@ -28,7 +32,7 @@ import (
 )
 
 const (
-	configMicroservices = "config-microservices.yaml"
+	configMicroservices = "config-microservices.tmpl.yaml"
 	configServerless    = "config-serverless.yaml"
 	configHA            = "config-scalable-single-binary.yaml"
 
@@ -131,122 +135,187 @@ func TestAllInOne(t *testing.T) {
 	}
 }
 
-func TestMicroservices(t *testing.T) {
-	s, err := e2e.NewScenario("tempo_e2e")
-	require.NoError(t, err)
-	defer s.Close()
-
-	minio := e2e_db.NewMinio(9000, "tempo")
-	require.NotNil(t, minio)
-	require.NoError(t, s.StartAndWaitReady(minio))
-
-	require.NoError(t, util.CopyFileToSharedDir(s, configMicroservices, "config.yaml"))
-	tempoIngester1 := util.NewTempoIngester(1)
-	tempoIngester2 := util.NewTempoIngester(2)
-	tempoIngester3 := util.NewTempoIngester(3)
-
-	tempoDistributor := util.NewTempoDistributor()
-	tempoQueryFrontend := util.NewTempoQueryFrontend()
-	tempoQuerier := util.NewTempoQuerier()
-	require.NoError(t, s.StartAndWaitReady(tempoIngester1, tempoIngester2, tempoIngester3, tempoDistributor, tempoQueryFrontend, tempoQuerier))
-
-	// wait for active ingesters
-	time.Sleep(1 * time.Second)
-	matchers := []*labels.Matcher{
+func TestMicroservicesWithKVStores(t *testing.T) {
+	testKVStores := []struct {
+		name     string
+		kvconfig func(port int) string
+	}{
 		{
-			Type:  labels.MatchEqual,
-			Name:  "name",
-			Value: "ingester",
+			name: "memberlist",
+			kvconfig: func(int) string {
+				return `
+		store: memberlist`
+			},
 		},
 		{
-			Type:  labels.MatchEqual,
-			Name:  "state",
-			Value: "ACTIVE",
+			name: "etcd",
+			kvconfig: func(port int) string {
+				return fmt.Sprintf(`
+		store: etcd
+		etcd:
+		  endpoints:
+		    - http://tempo_e2e-etcd:%d`, port)
+			},
+		},
+		{
+			name: "consul",
+			kvconfig: func(port int) string {
+				return fmt.Sprintf(`
+        store: consul
+        consul:
+          host: http://tempo_e2e-consul:%d`, port)
+			},
 		},
 	}
-	require.NoError(t, tempoDistributor.WaitSumMetricsWithOptions(e2e.Equals(3), []string{`cortex_ring_members`}, e2e.WithLabelMatchers(matchers...), e2e.WaitMissingMetrics))
 
-	// Get port for the Jaeger gRPC receiver endpoint
-	c, err := newJaegerGRPCClient(tempoDistributor.Endpoint(14250))
-	require.NoError(t, err)
-	require.NotNil(t, c)
+	for _, tc := range testKVStores {
+		t.Run(tc.name, func(t *testing.T) {
+			s, err := e2e.NewScenario("tempo_e2e")
+			require.NoError(t, err)
+			defer s.Close()
 
-	info := tempoUtil.NewTraceInfo(time.Now(), "")
-	require.NoError(t, info.EmitAllBatches(c))
+			// Set up KVStore
+			var kvstore *e2e.HTTPService
+			switch tc.name {
+			case "etcd":
+				kvstore = e2edb.NewETCD()
+				require.NoError(t, s.StartAndWaitReady(kvstore))
+			case "consul":
+				kvstore = e2edb.NewConsul()
+				require.NoError(t, s.StartAndWaitReady(kvstore))
+			case "memeberlist":
+			default:
+				t.Errorf("unknown KVStore %s", tc.name)
+			}
 
-	expected, err := info.ConstructTraceFromEpoch()
-	require.NoError(t, err)
+			tmpl, err := template.New(filepath.Base(configMicroservices)).ParseFiles(configMicroservices)
+			require.NoError(t, err)
 
-	// test metrics
-	require.NoError(t, tempoDistributor.WaitSumMetrics(e2e.Equals(spanCount(expected)), "tempo_distributor_spans_received_total"))
+			var KVStoreConfig string
+			if kvstore != nil {
+				KVStoreConfig = tc.kvconfig(kvstore.HTTPPort())
+			}
 
-	// test echo
-	assertEcho(t, "http://"+tempoQueryFrontend.Endpoint(3200)+"/api/echo")
+			var buf bytes.Buffer
+			kvconfig := map[string]interface{}{
+				"KVStore": KVStoreConfig,
+			}
+			require.NoError(t, tmpl.Execute(&buf, kvconfig))
 
-	// ensure trace is created in ingester (trace_idle_time has passed)
-	require.NoError(t, tempoIngester1.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_traces_created_total"))
-	require.NoError(t, tempoIngester2.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_traces_created_total"))
-	require.NoError(t, tempoIngester3.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_traces_created_total"))
+			require.NoError(t, util.WriteFileToSharedDir(s, "config.yaml", buf.Bytes()))
 
-	apiClient := tempoUtil.NewClient("http://"+tempoQueryFrontend.Endpoint(3200), "")
+			minio := e2edb.NewMinio(9000, "tempo")
+			require.NotNil(t, minio)
+			require.NoError(t, s.StartAndWaitReady(minio))
 
-	// query an in-memory trace
-	queryAndAssertTrace(t, apiClient, info)
+			tempoIngester1 := util.NewTempoIngester(1)
+			tempoIngester2 := util.NewTempoIngester(2)
+			tempoIngester3 := util.NewTempoIngester(3)
 
-	// search an in-memory trace
-	searchAndAssertTrace(t, apiClient, info)
+			tempoDistributor := util.NewTempoDistributor()
+			tempoQueryFrontend := util.NewTempoQueryFrontend()
+			tempoQuerier := util.NewTempoQuerier()
+			require.NoError(t, s.StartAndWaitReady(tempoIngester1, tempoIngester2, tempoIngester3, tempoDistributor, tempoQueryFrontend, tempoQuerier))
 
-	// flush trace to backend
-	res, err := e2e.DoGet("http://" + tempoIngester1.Endpoint(3200) + "/flush")
-	require.NoError(t, err)
-	require.Equal(t, 204, res.StatusCode)
+			// wait for active ingesters
+			time.Sleep(1 * time.Second)
+			matchers := []*labels.Matcher{
+				{
+					Type:  labels.MatchEqual,
+					Name:  "name",
+					Value: "ingester",
+				},
+				{
+					Type:  labels.MatchEqual,
+					Name:  "state",
+					Value: "ACTIVE",
+				},
+			}
+			require.NoError(t, tempoDistributor.WaitSumMetricsWithOptions(e2e.Equals(3), []string{`cortex_ring_members`}, e2e.WithLabelMatchers(matchers...), e2e.WaitMissingMetrics))
 
-	res, err = e2e.DoGet("http://" + tempoIngester2.Endpoint(3200) + "/flush")
-	require.NoError(t, err)
-	require.Equal(t, 204, res.StatusCode)
+			// Get port for the Jaeger gRPC receiver endpoint
+			c, err := newJaegerGRPCClient(tempoDistributor.Endpoint(14250))
+			require.NoError(t, err)
+			require.NotNil(t, c)
 
-	res, err = e2e.DoGet("http://" + tempoIngester3.Endpoint(3200) + "/flush")
-	require.NoError(t, err)
-	require.Equal(t, 204, res.StatusCode)
+			info := tempoUtil.NewTraceInfo(time.Now(), "")
+			require.NoError(t, info.EmitAllBatches(c))
 
-	// sleep for one maintenance cycle
-	time.Sleep(5 * time.Second)
+			expected, err := info.ConstructTraceFromEpoch()
+			require.NoError(t, err)
 
-	// test metrics
-	for _, i := range []*e2e.HTTPService{tempoIngester1, tempoIngester2, tempoIngester3} {
-		require.NoError(t, i.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_blocks_flushed_total"))
+			// test metrics
+			require.NoError(t, tempoDistributor.WaitSumMetrics(e2e.Equals(spanCount(expected)), "tempo_distributor_spans_received_total"))
+
+			// test echo
+			assertEcho(t, "http://"+tempoQueryFrontend.Endpoint(3200)+"/api/echo")
+
+			// ensure trace is created in ingester (trace_idle_time has passed)
+			require.NoError(t, tempoIngester1.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_traces_created_total"))
+			require.NoError(t, tempoIngester2.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_traces_created_total"))
+			require.NoError(t, tempoIngester3.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_traces_created_total"))
+
+			apiClient := tempoUtil.NewClient("http://"+tempoQueryFrontend.Endpoint(3200), "")
+
+			// query an in-memory trace
+			queryAndAssertTrace(t, apiClient, info)
+
+			// search an in-memory trace
+			searchAndAssertTrace(t, apiClient, info)
+
+			// flush trace to backend
+			res, err := e2e.DoGet("http://" + tempoIngester1.Endpoint(3200) + "/flush")
+			require.NoError(t, err)
+			require.Equal(t, 204, res.StatusCode)
+
+			res, err = e2e.DoGet("http://" + tempoIngester2.Endpoint(3200) + "/flush")
+			require.NoError(t, err)
+			require.Equal(t, 204, res.StatusCode)
+
+			res, err = e2e.DoGet("http://" + tempoIngester3.Endpoint(3200) + "/flush")
+			require.NoError(t, err)
+			require.Equal(t, 204, res.StatusCode)
+
+			// sleep for one maintenance cycle
+			time.Sleep(5 * time.Second)
+
+			// test metrics
+			for _, i := range []*e2e.HTTPService{tempoIngester1, tempoIngester2, tempoIngester3} {
+				require.NoError(t, i.WaitSumMetrics(e2e.Equals(1), "tempo_ingester_blocks_flushed_total"))
+			}
+			require.NoError(t, tempoQuerier.WaitSumMetrics(e2e.Equals(3), "tempodb_blocklist_length"))
+			require.NoError(t, tempoQueryFrontend.WaitSumMetrics(e2e.Equals(4), "tempo_query_frontend_queries_total"))
+
+			// query trace - should fetch from backend
+			queryAndAssertTrace(t, apiClient, info)
+
+			// stop an ingester and confirm we can still write and query
+			err = tempoIngester2.Kill()
+			require.NoError(t, err)
+
+			// sleep for heartbeat timeout
+			time.Sleep(1 * time.Second)
+
+			info = tempoUtil.NewTraceInfo(time.Now(), "")
+			require.NoError(t, info.EmitAllBatches(c))
+
+			// query an in-memory trace
+			queryAndAssertTrace(t, apiClient, info)
+
+			// search an in-memory trace
+			searchAndAssertTrace(t, apiClient, info)
+
+			// search the backend. this works b/c we're passing a start/end AND setting query ingesters within min/max to 0
+			now := time.Now()
+			searchAndAssertTraceBackend(t, apiClient, info, now.Add(-20*time.Minute).Unix(), now.Unix())
+
+			// stop another ingester and confirm things fail
+			err = tempoIngester1.Kill()
+			require.NoError(t, err)
+
+			require.Error(t, info.EmitBatches(c))
+		})
 	}
-	require.NoError(t, tempoQuerier.WaitSumMetrics(e2e.Equals(3), "tempodb_blocklist_length"))
-	require.NoError(t, tempoQueryFrontend.WaitSumMetrics(e2e.Equals(4), "tempo_query_frontend_queries_total"))
-
-	// query trace - should fetch from backend
-	queryAndAssertTrace(t, apiClient, info)
-
-	// stop an ingester and confirm we can still write and query
-	err = tempoIngester2.Kill()
-	require.NoError(t, err)
-
-	// sleep for heartbeat timeout
-	time.Sleep(1 * time.Second)
-
-	info = tempoUtil.NewTraceInfo(time.Now(), "")
-	require.NoError(t, info.EmitAllBatches(c))
-
-	// query an in-memory trace
-	queryAndAssertTrace(t, apiClient, info)
-
-	// search an in-memory trace
-	searchAndAssertTrace(t, apiClient, info)
-
-	// search the backend. this works b/c we're passing a start/end AND setting query ingesters within min/max to 0
-	now := time.Now()
-	searchAndAssertTraceBackend(t, apiClient, info, now.Add(-20*time.Minute).Unix(), now.Unix())
-
-	// stop another ingester and confirm things fail
-	err = tempoIngester1.Kill()
-	require.NoError(t, err)
-
-	require.Error(t, info.EmitBatches(c))
 }
 
 func TestScalableSingleBinary(t *testing.T) {
@@ -254,7 +323,7 @@ func TestScalableSingleBinary(t *testing.T) {
 	require.NoError(t, err)
 	defer s.Close()
 
-	minio := e2e_db.NewMinio(9000, "tempo")
+	minio := e2edb.NewMinio(9000, "tempo")
 	require.NotNil(t, minio)
 	require.NoError(t, s.StartAndWaitReady(minio))
 

--- a/integration/e2e/e2e_test.go
+++ b/integration/e2e/e2e_test.go
@@ -138,32 +138,32 @@ func TestAllInOne(t *testing.T) {
 func TestMicroservicesWithKVStores(t *testing.T) {
 	testKVStores := []struct {
 		name     string
-		kvconfig func(port int) string
+		kvconfig func(hostname string, port int) string
 	}{
 		{
 			name: "memberlist",
-			kvconfig: func(int) string {
+			kvconfig: func(string, int) string {
 				return `
-		store: memberlist`
+        store: memberlist`
 			},
 		},
 		{
 			name: "etcd",
-			kvconfig: func(port int) string {
+			kvconfig: func(hostname string, port int) string {
 				return fmt.Sprintf(`
-		store: etcd
-		etcd:
-		  endpoints:
-		    - http://tempo_e2e-etcd:%d`, port)
+        store: etcd
+        etcd:
+          endpoints:
+            - http://tempo_e2e-%s:%d`, hostname, port)
 			},
 		},
 		{
 			name: "consul",
-			kvconfig: func(port int) string {
+			kvconfig: func(hostname string, port int) string {
 				return fmt.Sprintf(`
         store: consul
         consul:
-          host: http://tempo_e2e-consul:%d`, port)
+          host: http://tempo_e2e-%s:%d`, hostname, port)
 			},
 		},
 	}
@@ -191,9 +191,9 @@ func TestMicroservicesWithKVStores(t *testing.T) {
 			tmpl, err := template.New(filepath.Base(configMicroservices)).ParseFiles(configMicroservices)
 			require.NoError(t, err)
 
-			var KVStoreConfig string
+			KVStoreConfig := tc.kvconfig("", 0)
 			if kvstore != nil {
-				KVStoreConfig = tc.kvconfig(kvstore.HTTPPort())
+				KVStoreConfig = tc.kvconfig(kvstore.Name(), kvstore.HTTPPort())
 			}
 
 			var buf bytes.Buffer

--- a/integration/e2e/e2e_test.go
+++ b/integration/e2e/e2e_test.go
@@ -183,7 +183,7 @@ func TestMicroservicesWithKVStores(t *testing.T) {
 			case "consul":
 				kvstore = e2edb.NewConsul()
 				require.NoError(t, s.StartAndWaitReady(kvstore))
-			case "memeberlist":
+			case "memberlist":
 			default:
 				t.Errorf("unknown KVStore %s", tc.name)
 			}

--- a/integration/e2e/e2e_test.go
+++ b/integration/e2e/e2e_test.go
@@ -154,7 +154,7 @@ func TestMicroservicesWithKVStores(t *testing.T) {
         store: etcd
         etcd:
           endpoints:
-            - http://tempo_e2e-%s:%d`, hostname, port)
+            - http://%s:%d`, hostname, port)
 			},
 		},
 		{
@@ -163,7 +163,7 @@ func TestMicroservicesWithKVStores(t *testing.T) {
 				return fmt.Sprintf(`
         store: consul
         consul:
-          host: http://tempo_e2e-%s:%d`, hostname, port)
+          host: http://%s:%d`, hostname, port)
 			},
 		},
 	}

--- a/pkg/gogocodec/gogocodec.go
+++ b/pkg/gogocodec/gogocodec.go
@@ -23,14 +23,15 @@ const (
 	jaegerProtoGenPkgPath = "github.com/jaegertracing/jaeger/proto-gen"
 	jaegerModelPkgPath    = "github.com/jaegertracing/jaeger/model"
 	otelProtoPkgPath      = "go.opentelemetry.io/collector"
-	etcdProtoPkgPath      = "go.etcd.io/etcd/api/v3/etcdserverpb"
+	// etcd path can be removed once upgrade to grpc >v1.38 is released (tentatively next release from v3.5.1)
+	etcdAPIProtoPkgPath = "go.etcd.io/etcd/api/v3"
 )
 
 func init() {
 	encoding.RegisterCodec(newCodec())
 }
 
-// gogoCodec forces the use of gogo proto marshalling/unmarshalling for Tempo/Cortex/Jaeger structs
+// gogoCodec forces the use of gogo proto marshalling/unmarshalling for Tempo/Cortex/Jaeger/etcd structs
 type gogoCodec struct {
 }
 
@@ -49,7 +50,7 @@ func (c *gogoCodec) Name() string {
 func (c *gogoCodec) Marshal(v interface{}) ([]byte, error) {
 	t := reflect.TypeOf(v)
 	elem := t.Elem()
-	// use gogo proto only for Tempo/Cortex/Jaeger types
+	// use gogo proto only for Tempo/Cortex/Jaeger/etcd types
 	if useGogo(elem) {
 		return gogoproto.Marshal(v.(gogoproto.Message))
 	}
@@ -60,18 +61,18 @@ func (c *gogoCodec) Marshal(v interface{}) ([]byte, error) {
 func (c *gogoCodec) Unmarshal(data []byte, v interface{}) error {
 	t := reflect.TypeOf(v)
 	elem := t.Elem()
-	// use gogo proto only for Tempo/Cortex/Jaeger types
+	// use gogo proto only for Tempo/Cortex/Jaeger/etcd types
 	if useGogo(elem) {
 		return gogoproto.Unmarshal(data, v.(gogoproto.Message))
 	}
 	return proto.Unmarshal(data, v.(proto.Message))
 }
 
-// useGogo checks if the element belongs to Tempo/Cortex/Jaeger packages
+// useGogo checks if the element belongs to Tempo/Cortex/Jaeger/etcd packages
 func useGogo(t reflect.Type) bool {
 	if t == nil {
 		return false
 	}
 	pkgPath := t.PkgPath()
-	return strings.HasPrefix(pkgPath, tempoProtoGenPkgPath) || strings.HasPrefix(pkgPath, cortexPath) || strings.HasPrefix(pkgPath, jaegerProtoGenPkgPath) || strings.HasPrefix(pkgPath, jaegerModelPkgPath) || strings.HasPrefix(pkgPath, otelProtoPkgPath) || strings.HasPrefix(pkgPath, etcdProtoPkgPath)
+	return strings.HasPrefix(pkgPath, tempoProtoGenPkgPath) || strings.HasPrefix(pkgPath, cortexPath) || strings.HasPrefix(pkgPath, jaegerProtoGenPkgPath) || strings.HasPrefix(pkgPath, jaegerModelPkgPath) || strings.HasPrefix(pkgPath, otelProtoPkgPath) || strings.HasPrefix(pkgPath, etcdAPIProtoPkgPath)
 }

--- a/pkg/gogocodec/gogocodec.go
+++ b/pkg/gogocodec/gogocodec.go
@@ -23,6 +23,7 @@ const (
 	jaegerProtoGenPkgPath = "github.com/jaegertracing/jaeger/proto-gen"
 	jaegerModelPkgPath    = "github.com/jaegertracing/jaeger/model"
 	otelProtoPkgPath      = "go.opentelemetry.io/collector"
+	etcdProtoPkgPath      = "go.etcd.io/etcd/api/v3/etcdserverpb"
 )
 
 func init() {
@@ -72,5 +73,5 @@ func useGogo(t reflect.Type) bool {
 		return false
 	}
 	pkgPath := t.PkgPath()
-	return strings.HasPrefix(pkgPath, tempoProtoGenPkgPath) || strings.HasPrefix(pkgPath, cortexPath) || strings.HasPrefix(pkgPath, jaegerProtoGenPkgPath) || strings.HasPrefix(pkgPath, jaegerModelPkgPath) || strings.HasPrefix(pkgPath, otelProtoPkgPath)
+	return strings.HasPrefix(pkgPath, tempoProtoGenPkgPath) || strings.HasPrefix(pkgPath, cortexPath) || strings.HasPrefix(pkgPath, jaegerProtoGenPkgPath) || strings.HasPrefix(pkgPath, jaegerModelPkgPath) || strings.HasPrefix(pkgPath, otelProtoPkgPath) || strings.HasPrefix(pkgPath, etcdProtoPkgPath)
 }

